etor/src/main/java/gov/hhs/cdc/trustedintermediary/etor/ruleengine/transformation/custom/MapLocalObservationCodes.java

```diff
@@ -0,0 +1,152 @@
+package gov.hhs.cdc.trustedintermediary.etor.ruleengine.transformation.custom;
+
+import gov.hhs.cdc.trustedintermediary.context.ApplicationContext;
+import gov.hhs.cdc.trustedintermediary.etor.messages.IdentifierCode;
+import gov.hhs.cdc.trustedintermediary.etor.ruleengine.FhirResource;
+import gov.hhs.cdc.trustedintermediary.etor.ruleengine.transformation.CustomFhirTransformation;
+import gov.hhs.cdc.trustedintermediary.external.hapi.HapiHelper;
+import gov.hhs.cdc.trustedintermediary.wrappers.Logger;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import org.hl7.fhir.r4.model.Bundle;
+import org.hl7.fhir.r4.model.Coding;
+import org.hl7.fhir.r4.model.Observation;
+import org.hl7.fhir.r4.model.StringType;
+
+/**
+ * Maps local observation codes to LOINC/PLT codes. In order to map a code, the observation must
+ * have a single local code with cwe-coding extension 'alt-coding'. The mapped LOINC or PLT code is
+ * added to the observation as an additional observation coding with cwe-coding extension 'coding'.
+ * When converted to/from HL7, these codings correspond to OBX-3.4/5/6 for the local code, and
+ * OBX-3.1/2/3 for the LOINC/PLT code.
+ */
+public class MapLocalObservationCodes implements CustomFhirTransformation {
+    protected final Logger LOGGER = ApplicationContext.getImplementation(Logger.class);
+
+    private HashMap<String, IdentifierCode> codingMap;
+
+    public MapLocalObservationCodes() {
+        initMap();
+    }
+
+    @Override
+    public void transform(FhirResource<?> resource, Map<String, String> args) {
```

Comment:
Are there more opportunities here to refactor and move code to HapiHelper? So we can reuse it for other transformations in the future?

Reply:
Great call and thanks for the suggestion. A lot of the logic is now in the HapiHelper class so it can be reused.